regidx.c

```diff
@@ -244,48 +329,136 @@ void regidx_destroy(regidx_t *idx)
+                size_t iori = ptr[i] - list->reg;
+                tmp_reg[i] = list->reg[iori];
+            }
+            free(ptr);
+            free(list->reg);
+            list->reg  = tmp_reg;
+            list->mreg = list->nreg;
+        }
+        list->unsorted = 0;
+    }
 
-    int iseq;
-    if ( khash_str2int_get(idx->seq2regs, chr, &iseq)!=0 ) return 0; // no such sequence
+    list->nidx = 0;
+    int j,k, midx = 0;
+    for (j=0; j<list->nreg; j++) {
+        int ibeg = iBIN(list->reg[j].beg);
+        int iend = iBIN(list->reg[j].end);
+        if ( midx <= iend )
+        {
+            int old_midx = midx;
+            midx = iend + 1;
+            kroundup32(midx);
+            uint32_t *new_idx = realloc(list->idx, midx*sizeof(uint32_t));
+            if (!new_idx) return -1;
+            list->idx = new_idx;
+            memset(list->idx+old_midx, 0, sizeof(uint32_t)*(midx-old_midx));
+        }
+        if ( ibeg==iend ) {
+            if ( !list->idx[ibeg] ) list->idx[ibeg] = j + 1;
+        } else {
+            for (k=ibeg; k<=iend; k++)
+                if ( !list->idx[k] ) list->idx[k] = j + 1;
+        }
+        if ( list->nidx < iend+1 ) list->nidx = iend+1;
+    }
+
+    return 0;
+}
+
+int regidx_overlap(regidx_t *regidx, const char *chr, hts_pos_t beg, hts_pos_t end, regitr_t *regitr)
+{
+    if ( regitr ) regitr->seq = NULL;
+
+    int iseq, ireg;
+    if ( khash_str2int_get(regidx->seq2regs, chr, &iseq)!=0 ) return 0;    // no such sequence
 
-    reglist_t *list = &idx->seq[iseq];
-    if ( !list->nregs ) return 0;
+    reglist_t *list = &regidx->seq[iseq];
+    if ( !list->nreg ) return 0;
 
-    int i, ibeg = from>>LIDX_SHIFT;
-    int ireg = ibeg < list->nidx ? list->idx[ibeg] : list->idx[ list->nidx - 1 ];
-    if ( ireg < 0 )
+    if ( list->nreg==1 )
     {
-        // linear search; if slow, replace with binary search
-        if ( ibeg > list->nidx ) ibeg = list->nidx;
-        for (i=ibeg - 1; i>=0; i--)
-            if ( list->idx[i] >=0 ) break;
-        ireg = i>=0 ? list->idx[i] : 0;
+        if ( beg > list->reg[0].end ) return 0;
+        if ( end < list->reg[0].beg ) return 0;
+        ireg = 0;
     }
-    for (i=ireg; i<list->nregs; i++)
+    else
     {
-        if ( list->regs[i].start > to ) return 0;   // no match
-        if ( list->regs[i].end >= from && list->regs[i].start <= to ) break; // found
+        if ( !list->idx ) {
+            if (reglist_build_index_(regidx,list) < 0) return -1;
+        }
+
+        int ibeg = iBIN(beg);
+        if ( ibeg >= list->nidx ) return 0;     // beg is too big
+
+        // find a matching region
+        uint32_t i = list->idx[ibeg];
+        if ( !i )
+        {
+            int iend = iBIN(end);
+            if ( iend > list->nidx ) iend = list->nidx;
+            for (i=ibeg; i<=iend; i++)
+                if ( list->idx[i] ) break;
+            if ( i>iend ) return 0;
+            i = list->idx[i];
+        }
```

Comment:
Similar to @pd3's comment above,  `ibeg` and `iend` come from `iBIN` so should be `hts_pos_t` too, as should that `uint32_t i` too.

PS.  For what it's worth, 31-bits + 13 bits shift equates to 16 GB before overflow.  That's conceivable for chromosome length on some organisms if I recall.

Reply:
I think you may be out by a factor of 1024 or so.  `2^(31+13) = 17592186044416`, so it should be good for up to 17 Tbases.  This limit is [here in the header file](https://github.com/samtools/htslib/blob/7ab4cac/htslib/regidx.h#L82) and is enforced in `regidx_push` so it shouldn't be possible to exceed this.

I think 17Tbases is likely to be enough for now.  We will be forced to amend it should anything be discovered that needs a bigger number.

Reply:
Oops yes you're right.  I miscounted my units.

In that case agreed.  It's not a practical concern, and even if it ever became one it's internal to the code and not an API/ABI problem.

Reply:
The maximum coordinate 1<<44 should be probably lowered. As it is now, keeping an index like this could require more than 8*((1<<44)>>13)/1e9 = 17GB. One of the largest known genomes (Picea abies) has 20e9 bp. Even if this was in a single chromosome, 1<<35 would suffice while keeping the memory in the order of tens of MB.

Reply:
The actual amount of memory used depends on the highest position in the region list, so normally it shouldn't be too big.  However there is a potential for overflow on 32-bit platforms when calculating the size of the memory buffer, so we probably do need to cut it down a bit more.  1 Tbp should be plenty big enough, even if the estimated genome size of [_Polychaos dubium_](https://www.ncbi.nlm.nih.gov/pubmed/16701223) turns out to be real.  That would give a maximum index size of 128Mb.